pkg/security/http/authz/oauth_scopes_test.go

```diff
@@ -1,51 +1,125 @@
-			claims:           `{"scope":["scope1","scope3"]}`,
-			expectError:      `none of the scopes [scope2] are present in the user token scopes [scope1 scope3]`,
-			expectedDecision: httpsec.Deny,
-			expectedAccess:   web.NoAccess,
-		}),
-		Entry("Succeeds if scope matches", testCase{
-			params:           []string{"scope2"},
-			claims:           `{"scope":["scope1","scope2","scope3"]}`,
-			expectedDecision: httpsec.Allow,
-			expectedAccess:   web.GlobalAccess,
-		}),
-	}...)
+var _ = Describe("Oauth Scopes", func() {
+	Describe("ScopeAuthorizer", func() {
+		DescribeTable("Required", func(t testCase) {
+			runTestCase(t, NewScopesAuthorizer(t.params.([]string), web.GlobalAccess))
+		}, []TableEntry{
+			Entry("Fails if no user is authenticated", testCase{
+				params:           []string{""},
+				noUser:           true,
+				expectedDecision: httpsec.Abstain,
+				expectedAccess:   web.NoAccess,
+			}),
+			Entry("Fails if token claims cannot be extracted", testCase{
+				params:           []string{""},
+				claimsError:      errors.New("claims error"),
+				expectError:      "could not extract scopes",
+				expectedDecision: httpsec.Deny,
+				expectedAccess:   web.NoAccess,
+			}),
+			Entry("Fails if there are no scopes in the token", testCase{
+				params:           []string{"scope2"},
+				claims:           `{}`,
+				expectError:      `none of the scopes [scope2] are present`,
+				expectedDecision: httpsec.Deny,
+				expectedAccess:   web.NoAccess,
+			}),
+			Entry("Fails if scope does not match", testCase{
+				params:           []string{"scope2"},
+				claims:           `{"scope":["scope1","scope3"]}`,
+				expectError:      `none of the scopes [scope2] are present in the user token scopes [scope1 scope3]`,
+				expectedDecision: httpsec.Deny,
+				expectedAccess:   web.NoAccess,
+			}),
+			Entry("Succeeds if scope matches", testCase{
+				params:           []string{"scope2"},
+				claims:           `{"scope":["scope1","scope2","scope3"]}`,
+				expectedDecision: httpsec.Allow,
+				expectedAccess:   web.GlobalAccess,
+			}),
+		}...)
+	})
+
+	Describe("HasScope", func() {
+		var (
+			tokenDataNoScopes              = `{"scope": []}`
+			tokenDataWithRequestedScope    = `{"scope": ["the-scope", "another-scope"]}`
+			tokenDataWithoutRequestedScope = `{"scope": ["another-scope"]}`
+		)
+
+		getUserContextWithToken := func(token string, err error) (*web.UserContext, error) {
+			ctx := web.ContextWithUser(context.Background(), &web.UserContext{
+				AuthenticationType: web.Bearer,
+				Data: func(data interface{}) error {
+					if err != nil {
+						return err
+					}
+					return json.Unmarshal([]byte(token), data)
+				},
+			})
+			user, ok := web.UserFromContext(ctx)
+			if !ok {
+				return nil, fmt.Errorf("Failed to retrieve user from context")
+			}
+
+			return user, nil
+		}
+
+		When("no scopes are available", func() {
+			user, err := getUserContextWithToken(tokenDataNoScopes, nil)
```

Comment:
Instead of writing this same line for every test, consider having `JustBeforeEach()` on top level which does this.
Then every test (inside every `When`) you can have a `BeforeEach()` which only sets up which token & err values to use in the `JustBeforeEach()`.

Also I see that every test starts with `Expect(err).ToNot(HaveOccurred())` - move that immediately after the `getUserContextWithToken()` invocation as that's where it makes more sense to be - closer to the actual function invocation itself, asserting that the completed function does not expect to have an error occurred.

Reply:
I tried to refactor to use Just/BeforeEach, but I think it turned out uglier.. it requires more test-scoped variables, and 4 lines of code to replace each single line I have now. 
I did refactor it a bit, for example removed the error checks for the user context creation as there are no errors expected there, can you have another look? If that's not the Go way of doing things I will change it, but I find the current implementation more readable than the BeforeEach() option.